README.md
CSV Generator
=============

[![Build Status](https://travis-ci.org/atomicpages/csv-generator.svg?branch=master)](https://travis-ci.org/atomicpages/csv-generator) [![Coverage Status](https://coveralls.io/repos/github/atomicpages/csv-generator/badge.svg?branch=master)](https://coveralls.io/github/atomicpages/csv-generator?branch=master) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

A simple-to-use script for generating CSV files with hundreds of columns and billions of rows.

### Installation
To install globally:

```bash
npm i -g csv-generator # or yarn
yarn global add csv-generator
```

Since the command is on our global path, we can execute via:

```bash
gencsv --help
```

you can also install locally:

```bash
npm i csv-generator
yarn add csv-generator
```

we can run locally:

```bash
./node_modules/.bin/gencsv --help
```

You can set a soft link to your project root also:

```bash
ln -s gencsv ./node_modules/.bin/gencsv
./gencsv --help
```

### Usage
You can use this script in three ways:

* Interactive CLI Interface
* CLI Interface
* API (used for embedding in your other scripts/modules)

#### Interactive Interface
The interactive interface is a series of questions and answers. This interface is not the default interface for this tool, but we can launch it via:

```bash
gencsv --interactive # or
gencsv -i
```

if you want always use the interactive interface, you can save this preference:

```bash
gencsv --always-interactive # or
gencsv -a
```

Don't need this interface anymore? No problem:

```bash
gencsv --clear-settings # or
gencsv -c
```

You can also run the script after clearing the settings if you wish:

```bash
gencsv -c foo.csv < columns.txt
```

#### CLI Interface
Resembling other commands like `ls`, this is the default feel:

```bash
encsv [options] <file> [columns..]
gencsv [options] <file> < columns.txt

Commands:
  gencsv <file>  The output file name

Options:
  --always-interactive, -a  Start the script in interactive mode and save this
                            setting
  --always-use-headers      Use this flag to persist column headers preferences
  --chunk                   The number of rows to generate per pass
                                                        [number] [default: 1000]
  --clear-settings, -c      Clear always-interactive settings
  --cluster                 Enable cluster mode       [boolean] [default: false]
  --cores                   The number of cores to use. Available cores: 8
                                                                        [number]
  --functions, --func       Lists available functions
  --interactive, -i         Run the script in interactive mode with a series of
                            questions and user-provided answers
  --partition               Manually set the partition size if you know what
                            works for your OS          [string] [default: "16K"]
  --rows, -r                The number of rows to generate (e.g. 100, 100000,
                            100k, 1M, 1B, etc.)                   [default: 100]
  --silent, -s              Minimal console output    [boolean] [default: false]
  --use-headers, -h         Use this flag to set column headers
                                                      [boolean] [default: false]
  --help                    Show help                                  [boolean]
  --version                 Show version number                        [boolean]
```

We can define our tables two ways:

1. Manual entry

    When issuing the command, be sure to separate functions by **space** &mdash; other delimiters are not supported with this method of entry.

    ```bash
    gencsv foo.csv '{{name.name}}, {{internet.email}}, {{finance.cc}}, {{date.recent}}, {{helpers.randomize(["a", "b"])}}'
    ```

    This will generate a `.csv` file in our current working directory with five columns and 100 rows.

    Need more rows? Need a lot of rows? No problem.

    ```bash
    gencsv foo.csv -r 100K '{{name.name}}, {{internet.email}}, {{finance.cc}}, {{date.format(null, "dd/mm/yyyy")}}' # or
    gencsv foo.csv -r 100000 '{{name.name}}, {{internet.email}}, {{finance.cc}}, {{date.format(null, "dd/mm/yyyy")}}'
    ```

    Same table, but this time with 100,000 rows!

    More? Go bananas:

    * `K` += 10^3 (e.g. 80K)
    * `M` += 10^6 (e.g. 0.2M)
    * `B` += 10^9 (e.g 2.13B)

    Need to interpolate shell variables? Honey badger don't care:

    ```bash
    gencsv foo.csv "{{name.name}}, {{internet.email}}, {{finance.cc}}, {{helpers.randomize(['$USER', 'John Doe', 'Jane Doe'])}}"
    ```

2. Columns definition as plain text

    For wide tables, we can define our column definitions in a separate file:

    ```bash
    gencsv foo.csv < columns.txt
    ```

    where `columns.txt` contains:

    ```
    {{name.name}}, {{internet.email}}, {{date.dob}}, {{random.ssn}}
    ```

    We can define column headers in `columns.txt` as well:

    ```
    Name, Email Address, Credit Card Number, Birthday, Transaction Date, ...
    {{name.name}}, {{internet.email}}, {{finance.cc}}, {{date.dob}}, {{date.recent}}, ...
    ```

#### Public CJS API
Building on top of this technology? Import the `csv-generator` module.

```js
const gencsv = require('csv-generator');

gencsv('foo.csv', ['name'], {
    rows: 100,
    chunks: 10,
    silent: true
}).then(
    res => {
        console.log(res);
    },
    e => {
        console.error(e);
    }
);
```

### Performance Notes
By nature, some functions are considerably slower than others, for example:

| Function | Time to Write 10K Rows by 100 Columns |
| :---: | :---: |
| `string` | 9.481s |
| `string(8)` | 6.305s |
| `alpha` | 6.138s |
| `alpha(8)` | 4.061s |
| `sentence` | 40.378s |

by contrast, these functions are much faster:

| Function | Time to Write 10K Rows by 100 Columns |
| :---: | :---: |
| `guid` | 1.264s |
| `age` | 0.72s |
| `integer` | 0.684s |
| `zip` | 2.579 |
| `yn` | 0.706s |

If you need to generate large amounts of data for _wide_ tables it is recommended to use fast functions.

#### More Notes
Instead of hundreds of async file writes (i.e. page faults), this generator uses a single stream to write content to the file. The trade-off is normal heap space, less CPU involvement, but more virtual memory used &mdash; memory is cheap; transistors aren't.

### Developers
Want to add a feature? Have a useful generator? Need to debug? The `readline` interface makes for cumbersome debugging, but node has our back:

```bash
node --inspect-brk index.js
```

or if you need to debug the CLI interface:

```bash
node --inspect-brk bin/gencsv foo.csv < columns.txt
```

open up Google Chrome and navigate to:

```
chrome://inspect
```

### Running Tests
To run unit tests:

```bash
npm run coverage # or
yarn coverage
```


CHANGELOG.md
### 2.0.0
* Setting minimum node version to align with LTS release
    * Updating travis build to reflect this
    * Adding node 9 and 10 to travis
    * Adding `yarn test` to travis scripts
* Upgrading dependencies
* General housekeeping
* Using faker for handle random data generation
* Deleting checked in dependencies and overhauling CSV generation
* Improving performance with file partitioning with `cluster`

### 1.0.3
* Bumping version for NPM

### 1.0.2
* Fixing clerical error in README
* Bumping node module versions
* Adding API docs
* Removing e2e tests since they're not complete
* Removing `asyncawait` plugin
* Disabling `dot-location` rule in ESLint
* Setting `ecmaVersion` to 2017 in ESLint
* using real `async/await` in `utils.js` now
* Exposing API publicly

    ```js
    const gencsv = require('csv-generator');

    gencsv('foo.csv', ['name'], {
        rows: 100,
        chunks: 10,
        silent: true
    }).then(
        res => {
            console.log(res);
        },
        e => {
            console.error(e);
        }
    );
    ```

### 1.0.1
* ~~Adding e2e tests~~
* Adding localizations
* Adding linting to travis build
* Fixing fatal exception when output file is missing for CLI interface
    * `gencsv -r 10K < columns.txt` displays a proper console error instead of a stacktrace
* Adding TONS of new benchmarks
    * `time` logs
    * `dtrace`
    * and spreadsheets with other cool stuff...
* Merging PR #1
	* Adding GUID
	* Removing float restrictions

### 1.0.0
* Initial Release


interactive.js
const chalk = require('chalk');
const figlet = require('figlet');
const inquirer = require('inquirer');
const Utils = require('./lib/utils');
const clear = require('clear');

/**
 * Creates the interactive interface where users can fill out the
 * details of their generated CSV via a question/answer model instead
 * of working with a config-based CLI.
 */
module.exports = function () {
    clear();

    console.log(''); // prevents top from being chopped off

    console.log(
        chalk.yellow(
            figlet.textSync('CSV Generator', {
                horizontalLayout: 'full'
            })
        )
    );

    console.log('To show help or list available functions, type "help" or "help functions" at any question and press ENTER');

    const help = input => {
        if (typeof input === 'string') {
            if (input.toLowerCase().trim() === 'help') {
                return 'Testing...';
            } else if (/^(help\s*)?func(tions?)?$/i.test(input.toLowerCase().trim())) {
                return 'Available functions:\n' + Utils.listFunctions(require('./functions'));
            }
        }

        return true;
    };

    inquirer.prompt([
        {
            name: 'file',
            type: 'input',
            message: 'Enter in the name of your output file:',
            default: 'my.csv',
            validate: function (input) {
                let response = help(input);

                if (response !== true) {
                    return response;
                }

                return true;
            }
        },
        {
            name: 'use-headers',
            type: 'list',
            message: 'Enable column headers?',
            choices: ['Yes', 'No']
        },
        {
            name: 'custom-headers',
            type: 'list',
            message: 'Specify custom column headers?',
            choices: ['Yes', 'No'],
            when: function (input) {
                return input['use-headers'] === 'Yes';
            }
        },
        {
            name: 'headers',
            message: 'Enter column headers separated by a comma.',
            type: 'editor',
            when: function (input) {
                return input['custom-headers'] === 'Yes';
            },
            validate: function (input) {
                let headers = input.trim().split(/\s*,\s*/);

                if (headers.length === 0) {
                    return 'Invalid column headers';
                }

                return true;
            }
        },
        {
            name: 'columns',
            type: 'input',
            message: 'Enter in column definitions using a list of available functions separated by a comma. For examples and other documentation, type "help" and press ENTER:',
            validate: function (input) {
                let response = help(input);

                if (response !== true) {
                    return response;
                }

                if (!input || (input && input.trim().split(/\s*,\s*/g).length === 0)) {
                    return 'Invalid column definitions.';
                }

                return true;
            }
        },
        {
            name: 'rows',
            type: 'input',
            message: 'Enter in the total number of rows to generate (e.g. 100, 100000, 100K, 10M, 1B, etc.):',
            default: 100,
            validate: function (input) {
                let response = help(input);

                if (response !== true) {
                    return response;
                }

                try {
                    Utils.convertNumber(input);
                } catch (e) {
                    return 'Invalid number entered. Value must be an integer greater than 0.';
                }

                return true;
            }
        }
    ]).then(answers => {
        Utils.generate(answers.file, answers.columns.split(/\s*,\s*/), {
            rows: Utils.convertNumber(answers.rows),
            chunks: 1000,
            headers: Array.isArray(answers.headers) ? answers.headers.split(/\s*,\s*/) : [])
        }
    });
};
